File: backend/app/chatbot.py
from fastapi import APIRouter, Body, Depends
from uuid import uuid4
import ollama
from app.mood_crisis_data import CRISIS_WORDS, DETECT_MOOD
from app.chat_history import store_chat, get_chat_history, create_new_conversation
from app.auth import get_current_user

router = APIRouter()

system_prompt = """
You are an empathetic mental health chatbot. Your role is to listen, support, and provide self-help strategies. 
You do NOT provide medical advice. If a user is in crisis, encourage them to seek professional help. Moreover, don't return your thinking part just return response.
"""

def get_chat_response(user_message, max_tokens=1000, temperature=0.7):
    messages = [{"role": "system", "content": system_prompt}, {"role": "user", "content": user_message}]
    response = ollama.chat(model="deepseek-r1", messages=messages, options={"max_tokens": max_tokens, "temperature": temperature})
    return response["message"]["content"]

def detect_mood(user_message):
    """
    Detects mood based on keywords in the user's message.
    DETECT_MOOD is a dictionary mapping mood keywords to a mood type.
    """
    for mood, keywords in DETECT_MOOD.items():
        if any(word in user_message.lower() for word in keywords):
            return mood
    return None  

@router.post("/")
async def chat(
    user_message: str = Body(..., embed=True),
    conversation_id: str = Body(None, embed=True),
    user_id: int = Depends(get_current_user)
):
    is_crisis = any(word in user_message.lower() for word in CRISIS_WORDS)
    mood = detect_mood(user_message) 

    if not conversation_id:
        conversation_id = str(uuid4())
        create_new_conversation(user_id, conversation_id)

    if is_crisis:
        response_text = "Please seek professional help. You're not alone ❤️."
    else:
        response_text = get_chat_response(user_message)

    store_chat(user_id, conversation_id, user_message, response_text, mood, is_crisis)

    return {"response": response_text, "conversation_id": conversation_id, "mood": mood}

@router.get("/history")
async def chat_history(limit: int = 10, user_id: int = Depends(get_current_user)):
    history = get_chat_history(user_id, limit)
    conversations = {}

    for row in history:
        conversation_id, user_message, bot_response, mood, is_crisis, timestamp = row
        if conversation_id not in conversations:
            conversations[conversation_id] = {
                "id": conversation_id,
                "messages": [],
                "timestamp": timestamp,
            }
        conversations[conversation_id]["messages"].append({
            "id": str(uuid4()),
            "user_message": user_message,
            "bot_response": bot_response,
            "mood": mood,
            "timestamp": timestamp,
            "is_crisis": is_crisis, 
        })
    
    return {"history": list(conversations.values())}

File: backend/app/chatbot.js
const express = require("express");
const router = express.Router();
const { v4: uuidv4 } = require("uuid");
const { getCurrentUser } = require("./auth");
const { CRISIS_WORDS, DETECT_MOOD } = require("./moodCrisisData");
const Conversation = require("./models/Conversation");
const Message = require("./models/Message");
const fetch = global.fetch; 

const getChatResponse = async (userMessage) => {
  // Use /api/generate and specify the model in the payload
  const url = process.env.MODEL_API_URL || "http://localhost:11434/api/generate";
  const payload = {
    model: "deepseek-r1:8b",
    prompt: userMessage,
    stream: false
  };
  try {
    const response = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload)
    });
    const text = await response.text();
    console.log("Raw response from model endpoint:", text);
    let data;
    try {
      data = JSON.parse(text);
    } catch (parseErr) {
      console.error("Failed to parse JSON. Raw response:", text);
      throw new Error("Failed to parse JSON from model response");
    }
    if (!response.ok) {
      console.error("Ollama responded with error:", data);
      throw new Error(data.error?.message || "Ollama API error");
    }
    return data.response ? data.response.trim() : data.result.trim();
  } catch (err) {
    console.error("Error in getChatResponse:", err);
    throw err;
  }
};

const detectMood = (userMessage) => {
  const lower = userMessage.toLowerCase();
  for (const [mood, keywords] of Object.entries(DETECT_MOOD)) {
    if (keywords.some((word) => lower.includes(word))) {
      return mood;
    }
  }
  return null;
};

const createNewConversation = async (userId, conversationId) => {
  const conversation = new Conversation({ _id: conversationId, user_id: userId });
  await conversation.save();
};

const storeChat = async (userId, conversationId, userMessage, botResponse, mood, isCrisis) => {
  const messageId = uuidv4();
  const message = new Message({
    _id: messageId,
    user_id: userId,
    conversation_id: conversationId,
    user_message: userMessage,
    bot_response: botResponse,
    mood,
    is_crisis: isCrisis,
  });
  await message.save();
};

router.post("/", getCurrentUser, async (req, res) => {
  const { user } = req;
  let { user_message, conversation_id } = req.body;
  const isCrisis = CRISIS_WORDS.some(word => user_message.toLowerCase().includes(word));
  const mood = detectMood(user_message);

  try {
    if (!conversation_id) {
      conversation_id = uuidv4();
      await createNewConversation(user.id, conversation_id);
    }
    let responseText;
    if (isCrisis)
      responseText = "Please seek professional help. You're not alone ❤️.";
    else {
      try {
        responseText = await getChatResponse(user_message);
      } catch (err) {
        return res.status(500).json({ detail: err.message });
      }
    }
    await storeChat(user.id, conversation_id, user_message, responseText, mood, isCrisis);
    res.json({ response: responseText, conversation_id, mood });
  } catch (err) {
    return res.status(500).json({ detail: err.message });
  }
});

// GET /chat/history
router.get("/history", getCurrentUser, async (req, res) => {
  const { user } = req;
  const limit = parseInt(req.query.limit) || 10;
  try {
    const messages = await Message.find({ user_id: user.id })
      .sort({ timestamp: -1 })
      .limit(limit)
      .lean();
    // Group messages by conversation_id
    const conversations = {};
    messages.forEach((row) => {
      if (!conversations[row.conversation_id])
        conversations[row.conversation_id] = { id: row.conversation_id, messages: [], timestamp: row.timestamp };
      conversations[row.conversation_id].messages.push({
        id: row._id,
        user_message: row.user_message,
        bot_response: row.bot_response,
        mood: row.mood,
        is_crisis: row.is_crisis,
        timestamp: row.timestamp
      });
    });
    res.json({ history: Object.values(conversations) });
  } catch (err) {
    return res.status(500).json({ detail: err.message });
  }
});

module.exports = { chatbotRouter: router };
